Fix changelog generator

littlejawa opened the pull request:
I've been testing the script on various situations, and fixing problems as I find them.
Hopefully, this should be usable again.

gkurz committed:
test-fbc: Handle pagination in quay.io API

Our pipeline is currently failing to create test catalogs. It
appears to be looping on :

  jq: error (at <stdin>:1): Cannot iterate over null (null)
  jq: error (at <stdin>:1): Cannot iterate over null (null)
  jq: error (at <stdin>:1): Cannot iterate over null (null)
  jq: error (at <stdin>:1): Cannot iterate over null (null)
  jq: error (at <stdin>:1): Cannot iterate over null (null)

This comes from a infinite loop when trying to generate the
changelog. Leverage `.has_additional` from the quay.io to
break out the loop.

Signed-off-by: Greg Kurz <groug@kaod.org>

littlejawa committed:
test-fbc: do not fail on changelog creation errors

Changelog creation issues should not prevent the build of the catalog image.
This commit makes sure that an error in the script won't break the build.
- always succeed when running changelog.sh
- retrieve the script's output in a separate file (CHANGELOG.err) for
  troubleshooting, and add this file to the resulting image
- make sure the CHANGELOG and CHANGELOG.err are created even if an error
  happens, or if the parameters for the git URL was not provided.

Signed-off-by: Julien Ropé <jrope@redhat.com>

littlejawa committed:
test-fbc: add parameters to the Dockerfile that allow custom changelog creation

This is not used in general, but can help troubleshooting the changelog creation.
The parameters can be used to generate a changelog for a specific set of changes.

Signed-off-by: Julien Ropé <jrope@redhat.com>

littlejawa committed:
test-fbc: avoid infinite loop in changelog.sh

When a problem occurs during the processing, catch the error and get out of the
loop with an error, rather than looping forever.

Signed-off-by: Julien Ropé <jrope@redhat.com>

littlejawa committed:
test-fbc: display commit message for changes that don't update the bundle image

Sometimes, commits that changes catalog-template.yaml are not changing the
bundle image reference.
Typically: bumping version, or merge commits.

For those, since there is no "new image" to look for, just display the commit
message for the catalog-template.yaml change.

Signed-off-by: Julien Ropé <jrope@redhat.com>

littlejawa committed:
test-fbc: fix images names for repository lookup

We were using Konflux components names, not the image names that are visible in
the CSV, causing errors when trying to get the changelog for those images.

Signed-off-by: Julien Ropé <jrope@redhat.com>

littlejawa committed:
test-fbc: allow changelog.sh to log image additions properly

When a new image is added, there is nothing to compare to, and the script fails.
This commit catches this situation and displays the list of new images, if there
is more than one.

Note that we still assume there is only one image *update* per commit in the
general case - if we find more than one change, we currently only log the changes
for the first one.
This is consistent with Konflux's current behavior of making one commit per
image change.

Signed-off-by: Julien Ropé <jrope@redhat.com>

gkurz commented on fbc/test-fbc/changelog.sh:
```diff
@@ -46,6 +46,10 @@ function get_commit_for_image() {
+        if [ $? -ne 0 ]; then
+            echo "Error: Failed to parse JSON response from quay.io for image $IMAGE_NAME" >&2
+            break
+        fi
```
FWIW I had identified the reason for the infinite loop as a lack of GH pagination handling (see https://github.com/gkurz/sandboxed-containers-operator/commit/993830782603ae0fbbe5aea5b39908e970707d29).

littlejawa replied:
I have added your commit too. It's cleaner to handle the pagination than to wait for an error to happen.
I keep the error check here anyway, just in case something else causes a problem.

littlejawa committed:
test-fbc: do not try to log changes for version string updates

When we update the version strings, we generate changes that can't be compared
to previous version in git.
Since we use comment markers for the strings to update, the marker can be used
in the script to catch this situation.
This commit makes sure we don't try to do an image changelog for version bumps.

Signed-off-by: Julien Ropé <jrope@redhat.com>

littlejawa committed:
test-fbc: filter our nudge PR commit from changelog

The changelog.sh script lists all changes to the bundle, i.e: all nudge PRs.
Having them listed additionally in every log from the operator's repo is redundant
and doen't bring any value to the changelog.

Signed-off-by: Julien Ropé <jrope@redhat.com>

gkurz commented on fbc/test-fbc/changelog.sh:
```diff
@@ -115,22 +115,36 @@ COMMIT_LIST=$(git log --oneline "${OLD_BUNDLE_COMMIT}..${NEW_BUNDLE_COMMIT}" ../
+    # take the first one, if there are more
+    OLD_IMAGE=$(echo "$OLD_IMAGES" | head -n1)
+    NEW_IMAGE=$(echo "$NEW_IMAGES" | head -n1)
+
```
Maybe add a sanity check to ensure NEW_IMAGE is indeed a new version of OLD_IMAGE ? Just in case, to avoid confusion.

littlejawa replied:
Discussed with Greg, putting our conclusion here:
This is nice to have, but should come on a later update. The reason is that this script is meant to be used in the CI, where there is 0 chance that we get bad input for commit references.
The situation may arise for local execution, and we certainly want to sanitize the inputs there, but it can come later.
First, let's get this script running in test-fbc builds, and get changelogs with our catalog.

gkurz commented on fbc/test-fbc/changelog.sh:
```diff
@@ -145,6 +145,16 @@ for COMMIT in $COMMIT_LIST; do
 
+    # Corner case: when we update the version strings, we get OSC_VERSION as the
+    # image reference. This is because we mark all version strings to update with
+    # the comment "# OSC_VERSION".
```
This is clever but it seems a bit hacky to rely on this implementation detail in the parsing of `image:` lines. What about an explicit check upfront to skip `image: .* ## OSC_VERSION` ?

littlejawa replied:
Right... Filtering out the lines with OSC_VERSION from the start is much cleaner actually. 
I just modified this commit (see c18c1b6)

gkurz commented on fbc/test-fbc/Dockerfile:
```diff
@@ -17,22 +17,21 @@ RUN cp -r test-fbc/catalog/ /configs
+        pushd repo && \
+        git checkout "${GIT_REVISION}" && \
+        cd fbc/test-fbc && \
+        ./changelog.sh 2>&1 | tee CHANGELOG.err || true && \
```
`2>&1` will result in CHANGELOG.err to also have the content of CHANGELOG since `changelog.sh` does `tee CHANGELOG` already. Is that intended ?

gkurz replied:
Yes it is otherwise it would be hard to guess which command failed :smiley:

littlejawa replied:
Yes, that's the intent.
Having only the errors in CHANGELOG.err makes it difficult to troubleshoot, because we don't know what the error messages relates to in the script's execution.

gkurz commented on fbc/test-fbc/Dockerfile:
```diff
@@ -3,6 +3,8 @@
 ARG GIT_URL=""
 ARG GIT_REVISION=""
+ARG OLDCOMMIT=""
+ARG NEWCOMMIT=""
```
This is for the case where one doesn't want the script directly in their repo, right ? If so, maybe worth to document this in some README.md ?

gkurz replied:
Not strictly needed.

littlejawa committed:
test-fbc: make it possible to call changelog.sh with arbirary commit ranges

It is cumbersome to verify what commits are changing the catalog-template.yaml
file. The test-fbc build is fine, as it is, by definition, running on such a
change. But when we invoke the script locally, we may want more flexibility.

This commit uses the range provided as parameter, and gets the first and last
commit in this range that impacts catalog-template.yaml.

Signed-off-by: Julien Ropé <jrope@redhat.com>

gkurz reviewed (approved):
Great work ! Thanks @littlejawa !

/lgtm
